src/core/geometry/qgsgeometry.h

```diff
@@ -788,6 +788,15 @@ class CORE_EXPORT QgsGeometry
      */
     bool insertVertex( const QgsPoint &point, int beforeVertex );
 
+    /**
+     * Adds a vertex to the segment which intersect \a point but don't
+     * already have a vertex there. Closest segment is identified using \a segmentSearchEpsilon.
+     * If a vertex already exists within \a snappingTolearnceDistance, no additional vertex is inserted.
+     * \returns TRUE if point was added, FALSE otherwise
+     * \since QGIS 3.38
+     */
+    bool addTopologicalPoint( const QgsPoint &point, double snappingTolerance = 1e-7, double segmentSearchEpsilon = 1e-12 );
```

Comment:
I've picked the worst case values from the current _add topological points_ implementation and reused those as the default values for the new method:
https://github.com/qgis/QGIS/blob/3da05d3fc93e52abf593505e4fc66f60a1ba09af/src/core/vector/qgsvectorlayereditutils.cpp#L783-L790

The method is always called with explicit parameters is QGIS code, so these are mostly a fail safe default for python users who might not know what values to pick.
I'm open to using a different set of values there if you think it would make more sense!

Reply:
OK, it makes sense for 1e-12, but, please, use 1e-8 for `snappingTolerance`. It's not for the -8 vs -7, just to be consistent elsewhere. Moreover, I'd like to propose a variable/settings to adjust this tolerance after https://github.com/qgis/QGIS/pull/55648

Thanks.